### PR TITLE
PE-1645: Query for Inferno data via chain data rather than GQL

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

**Comment by matibat** [comment]
TODO: track the `bundledIn` field

File: src/common.ts [hunk 1]
```diff
@@ -334,3 +334,9 @@ export function changeInPercentage(prev: number, curr: number): number {
 		return (curr - prev) / prev;
 	}
 }
+
+const initialErrorDelayMS = 1000;
+export async function exponentialBackOffAfterFailedRequest(retryNumber: number): Promise<void> {
```
Comment (arielmelendez) [inline src/common.ts]: It's better to use a library for this kind of thing (e.g. axios-retry).

File: src/queries.ts [hunk 2]
```diff
@@ -58,155 +57,31 @@ export async function getAllArDriveTransactionsWithin(range: HeightRange): Promi
...
 	return allEdges;
 }
-
-/**
- * Runs the given GQL query
- * @param query the query object
- * @throws if the GW returns a syntax error
- * @returns {GQLTransactionsResultInterface} the returned transactions
- */
-async function sendQuery(query: Query): Promise<GQLTransactionsResultInterface> {
```
Comment (arielmelendez) [inline src/queries.ts:98]: I don't think it was strictly necessary to remove all this. If GQL becomes viable again, we'd still want to use this.

File: src/queries_layer_1.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import fetch from 'node-fetch';
+import { exponentialBackOffAfterFailedRequest } from './common';
+import { MAX_RETRIES, WINSTON_AR_ASPECT } from './constants';
+import { GQLEdgeInterface, GQLTagInterface } from './gql_types';
+import { decodeTags, fromB64Url, sha256B64Url } from './utils/layer_one_helpers';
+
+export async function getAllParsedTransactionsOfBlock(height: number): Promise<GQLEdgeInterface[]> {
+	const block = await getBlock(height);
+	const txs = await getAllTransactionsOfBlock(block);
+	return txs.map(parseLayer1Tx.bind(undefined, block));
```
Comment (arielmelendez) [inline src/queries_layer_1.ts]: Why is this bind necessary?

File: src/queries_layer_1.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	while (!responseOk && pendingRetries >= 0) {
+		const currentRetry = MAX_RETRIES - pendingRetries;
+		if (pendingRetries !== MAX_RETRIES) {
+			await exponentialBackOffAfterFailedRequest(currentRetry);
+		}
+
+		try {
+			const txIDs = block.txs;
+			const requests = txIDs.map((txId) => fetch(`https://arweave.net/tx/${txId}`));
+			const responses = await Promise.all(requests);
```
Comment (arielmelendez) [inline src/queries_layer_1.ts]: This looks pretty fragile.

File: src/queries_layer_1.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+			const transactions: L1Transaction[] = await Promise.all(responses.map((resp) => resp.json()));
+			return transactions;
+		} catch (e) {
+			pendingRetries--;
```
Comment (arielmelendez) [inline src/queries_layer_1.ts]: Why retry ALL the transactions rather than just the ones we need?

File: src/queries_layer_1.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+			await exponentialBackOffAfterFailedRequest(currentRetry);
+		}
+
+		try {
+			const url = `https://arweave.net/block/height/${height}`;
+			const response = await fetch(url);
+			const block: L1Block = await response.json();
+			return block;
+		} catch (e) {
+			pendingRetries--;
```
Comment (arielmelendez) [inline src/queries_layer_1.ts]: Writing all this retry logic by hand is very error prone. We should just use a library.

File: src/common.ts (hunk 1 above)
Comment (matibat) [reply src/common.ts]: Actually this function came from axis-retry

File: src/queries.ts (hunk 2 above)
Comment (matibat) [reply src/queries.ts:98]: We can always go back in the gh history to get that code back

File: src/queries_layer_1.ts (hunk 3 above)
Comment (matibat) [reply src/queries_layer_1.ts]: TODO: use the factory pattern instead of this

File: src/queries_layer_1.ts (hunk 5 above)
Comment (matibat) [reply src/queries_layer_1.ts]: 🧠

File: src/queries_layer_1.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+	axiosRetry(axiosInstance, {
+		retries: MAX_RETRIES,
+		retryDelay: (retryNumber) => {
+			// 	console.error(`Retry attempt ${retryNumber}/${maxRetries} of request to ${reqURL}`);
+			return exponentialDelay(retryNumber);
+		}
+	});
+
+	const parallelize = pLimit.default(25);
```
Comment (matibat) [inline src/queries_layer_1.ts]: Clean this up: pLimit.default

File: src/queries_layer_1.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+				return axiosInstance.get(`https://arweave.net/tx/${txid}`);
+			});
+		})
+	);
+
+	responses.forEach((resp) => {
+		if (resp.status !== 200) {
+			console.log(`Response code (${resp.status}): ${resp.statusText}`);
+		}
+	});
```
Comment (matibat) [inline src/queries_layer_1.ts]: TODO: Log if it is a Request Limit err code

File: src/utils/layer_one_helpers.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+				value: fromB64Url(tag.value).toString()
+			};
+		} catch (error) {
+			return { name: 'undefined', value: 'undefined' };
+		}
+	});
+	return tags;
+}
+
+export function getTagValue(tags: GQLTagInterface[], name: string): string | undefined {
```
Comment (matibat) [inline src/utils/layer_one_helpers.ts]: Unused

**Review by matibat — COMMENTED** [review]
Rename the "layer_one_helpers.ts" to "layer_1_helpers.ts"

File: src/queries_layer_1.ts (hunk 8 above)
Comment (matibat) [reply src/queries_layer_1.ts]: We can remove this

File: src/queries_layer_1.ts (hunk 3 above)
Comment (matibat) [reply src/queries_layer_1.ts]: 2fa14eb8b4025a1a81c699069f98d318822c8cdd

File: src/queries_layer_1.ts (hunk 8 above)
Comment (matibat) [reply src/queries_layer_1.ts]: 2fa14eb8b4025a1a81c699069f98d318822c8cdd

**Review by karlprieb — APPROVED** [review]

**Pull request merged**